Doc/library/typing.rst

```diff
@@ -1266,213 +1323,327 @@ These can be used as types in annotations using ``[]``, each having a unique syn
           try:
               return mapping[key]
           except KeyError:
               return default
 
-.. class:: TypeVar
+   Here the brackets after the function name indicate a
+   :ref:`generic function <generic-functions>`.
 
-    Type variable.
+   For backwards compatibility reasons, generic classes can also be
+   declared by explicitly inheriting from
+   ``Generic``. In this case, the type parameters must be declared
+   separately::
 
-    Usage::
+      KT = TypeVar('KT')
+      VT = TypeVar('VT')
+
+      class Mapping(Generic[KT, VT]):
+          def __getitem__(self, key: KT) -> VT:
+              ...
+              # Etc.
+
+.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False)
+
+   Type variable.
+
+   The preferred way to construct a type variable is via the dedicated syntax
+   for :ref:`generic functions <generic-functions>`,
+   :ref:`generic classes <generic-classes>`, and
+   :ref:`generic type aliases <generic-type-aliases>`::
+
+      class Sequence[T]:  # T is a TypeVar
+          ...
+
+   This syntax can also be used to create bound and constrained type
+   variables::
+
+      class StrSequence[S: str]:  # S is a TypeVar bound to str
+          ...
+
+
+      class StrOrBytesSequence[A: (str, bytes)]:  # A is a TypeVar constrained to str or bytes
+          ...
+
+   However, if desired, reusable type variables can also be constructed manually, like so::
 
       T = TypeVar('T')  # Can be anything
       S = TypeVar('S', bound=str)  # Can be any subtype of str
       A = TypeVar('A', str, bytes)  # Must be exactly str or bytes
 
-    Type variables exist primarily for the benefit of static type
-    checkers.  They serve as the parameters for generic types as well
-    as for generic function definitions.  See :class:`Generic` for more
-    information on generic types.  Generic functions work as follows::
+   Type variables exist primarily for the benefit of static type
+   checkers.  They serve as the parameters for generic types as well
+   as for generic function and type alias definitions.
+   See :class:`Generic` for more
+   information on generic types.  Generic functions work as follows::
+
+      def repeat[T](x: T, n: int) -> Sequence[T]:
+          """Return a list containing n references to x."""
+          return [x]*n
+
+
+      def print_capitalized[S: str](x: S) -> S:
+          """Print x capitalized, and return x."""
+          print(x.capitalize())
+          return x
+
+
+      def concatenate[A: (str, bytes)](x: A, y: A) -> A:
+          """Add two strings or bytes objects together."""
+          return x + y
+
+   Note that type variables can be *bound*, *constrained*, or neither, but
+   cannot be both bound *and* constrained.
+
+   The variance of type variables is inferred by type checkers when they are created
+   through the :ref:`type parameter syntax <type-params>` and when
```

Comment:
```suggestion
   through the :ref:`type parameter syntax <type-params>` or when
```

Reply:
I see what @Gobot1234 means, actually; I think I have a slight preference for "or". But no strong opinion.